plugins/timertop/timermodel.cpp

```diff
@@ -27,190 +27,358 @@
+            auto it = s_gatheredTimersData.find(id);
+
+            if (it == s_gatheredTimersData.end()) {
+                it = s_gatheredTimersData.insert(id, TimerIdData());
+            }
+
+            const TimeoutEvent timeoutEvent(QTime::currentTime(), -1);
+
+            it.value().update(id, receiver);
+            it.value().addEvent(timeoutEvent);
+
+            s_timerModel->m_triggerPushChangesMethod.invoke(s_timerModel, Qt::QueuedConnection);
+        }
     }
 
-    const TimerInfoPtr timerInfo = timerInfoVariant.value<TimerInfoPtr>();
-    Q_ASSERT(timerInfo->timerObject() == timer);
-    return timerInfo;
+    return false;
 }
 
-TimerInfoPtr TimerModel::findOrCreateQTimerTimerInfo(int timerId)
+TimerModel::~TimerModel()
 {
-    for (int row = 0; row < m_sourceModel->rowCount(); row++) {
-        const QModelIndex sourceIndex = m_sourceModel->index(row, 0);
-        QObject * const timerObject = sourceIndex.data(ObjectModel::ObjectRole).value<QObject *>();
-        QTimer * const timer = qobject_cast<QTimer *>(timerObject);
-        if (timer && timer->timerId() == timerId)
-            return findOrCreateQTimerTimerInfo(timer);
-    }
-
-    return TimerInfoPtr();
+    QMutexLocker locker(&s_mutex);
+    QInternal::unregisterCallback(QInternal::EventNotifyCallback, eventNotifyCallback);
+    m_timersInfo.clear();
+    s_gatheredTimersData.clear();
 }
 
-TimerInfoPtr TimerModel::findOrCreateTimerInfo(const QModelIndex &index)
+bool TimerModel::isInitialized()
 {
-    if (index.row() < m_sourceModel->rowCount()) {
-        const QModelIndex sourceIndex = m_sourceModel->index(index.row(), 0);
-        QObject * const timerObject = sourceIndex.data(ObjectModel::ObjectRole).value<QObject *>();
-        return findOrCreateQTimerTimerInfo(timerObject);
-    } else {
-        const int freeListIndex = index.row() - m_sourceModel->rowCount();
-        Q_ASSERT(freeListIndex >= 0);
-        if (freeListIndex < m_freeTimers.size())
-            return m_freeTimers.at(freeListIndex);
-    }
-    return TimerInfoPtr();
+    return s_timerModel != nullptr;
 }
 
-int TimerModel::rowFor(QObject *timer)
+TimerModel *TimerModel::instance()
 {
-    for (int i = 0; i < rowCount(); i++) {
-        const TimerInfoPtr timerInfo = findOrCreateTimerInfo(index(i, 0));
-        if (timerInfo && timerInfo->timerObject() == timer)
-            return i;
-    }
-    return -1;
+    if (!s_timerModel)
+        s_timerModel = new TimerModel;
+
+    Q_ASSERT(s_timerModel);
+    return s_timerModel;
 }
 
 void TimerModel::preSignalActivate(QObject *caller, int methodIndex)
 {
-    if (!(methodIndex == m_timeoutIndex && qobject_cast<QTimer *>(caller))
-        && !(methodIndex == m_qmlTimerTriggeredIndex && caller->inherits("QQmlTimer")))
-        return;
+    // We are in the thread of the caller emmiting the signal
+    // The probe did NOT locked the objectLock at this point.
+    Q_ASSERT(TimerModel::isInitialized());
+
+    QMutexLocker locker(&s_mutex);
```

Comment:
Can we move this down to after the check if this is a relevant signal for us? Up to there this is performance-sensitive hot-path code (called for every single signal emission!), so we don't want to serialize this, nor add the cost of the mutex operation here if we can help it.